Response to #40

carlosjoserg opened the pull request:
Could be that #40 happened mainly due to 2 things (well, 3, if you count my awkwardness):

1. Lack of tests on command computation, as already requested by @adolfo-rt in #41
2. The `computeCommand` method has two signatures with almost a replica of code within -> Same code in two places requires update in two places -> The chance of making a mistake when updating increases.

In this initial PR both are addressed, there are 2 separated commits to easy the review.

First, in 9dc84fd548c0bd7df9d428cf60594fde24ea2b6d, the coverage of the PID class test suite is increased with the following modification of current tests:

* `zeroITermBadIBoundsTest`: I think that entering `i_max < i_min` should be prevented, or at least warned. A suggested change to avoid bad i-bounds is there commented because it would require a change in the `setGains` method return type to `bool` instead of `void` and the current implementation wouldn't pass the test. Another posibility would be to leave that return type and just throw a warning if `i_max < i_min` is passed.

* `integrationWindupTest`: Now upper and lower bounds are tested.

* integrationWindupZeroGainTest: Untouched. However, it looks to me it is useless with the current integral contribution implementation.

* `gainSettingCopyPIDTest`: Makeup in names for readability + Added test on `reset()` method.

And the following additions, a new set of tests under the name `CommandTest`, hopefully well described with comments:

* `proportionalOnlyTest`, `integralOnlyTest`, `derivativeOnlyTest`: Check that the command computed using each contribution independently is correct. Note that, they depend on the integral and differentiation schemes, and if they change, the tests might not be valid anymore.

* `completePIDTest`: Few checks on the sum up of contributions. Same here, the test depends on the integral and differentiation schemes.

Second, in 7fe5e126b05c4c915ecbb4c143ffef6b8c0ff5c1, only one method does the PID command computation and writes to class members (except for `p_error_last_`), which is the one with the signature that allows users to pass a pre-computed error derivative. The other one calls this method passing the error derivative computed with an internal differentiation scheme. I couldn't find whether this is ROS-style compliant or not, and not sure if it affects critically the performance, but surely it will reduce the chance of making a mistake if the implementation is to be updated again.

PS: I didn't do anything with the DEPRECATED command computation methods.

carlosjoserg commented:
To make issue 2 in this PR clearer, check that, in #38, only one of the `computeCommand` methods is updated. So, if someone uses the other signature, the added feature wouldn't be available.

paulbovbel commented:
+1, this supercedes #44, which can be closed if this is merged

adolfo-rt commented on test/pid_tests.cpp:
```diff
@@ -11,15 +11,22 @@ using namespace control_toolbox;
+  RecordProperty("description","This test checks that the limits for the integral contribution are entered correctly.");
 
-  Pid pid(1.0, 0.0, 0.0, -1.0, 0.0);
+  //// Do not allow this construction
```
Please either remove dead code, or bring it to a functional state.

adolfo-rt commented:
Could you please rebase against `indigo-devel`. I think it should be straightforward (or fast-forward, to be more precise ;-).

Many thanks for the extra tests.

carlosjoserg commented:
Sure. 

I'll also delete 7fe5e126b05c4c915ecbb4c143ffef6b8c0ff5c1 since #44 was already merged, it was exactly the same.

carlosjoserg commented:
[update]
- Rebased to `indigo-devel` (all tests passed)
- Minor, changed the name of `zeroITermBadIBoundsTest` to `ITermBadIBoundsTest`
- Deleted 7fe5e126b05c4c915ecbb4c143ffef6b8c0ff5c1 from PR

bmagyar commented:
I'm about to press some green buttons here. It seems that this PR got a bit forgotten, please raise a thumb if you also think it should be merged.

efernandez commented on test/pid_tests.cpp:
```diff
@@ -152,11 +153,143 @@ TEST(ParameterTest, gainSettingCopyPIDTest)
+
+  // If call again with same arguments
+  cmd = pid.computeCommand(-0.5, ros::Duration(1.0));
+  // Then expect the integration do it part doubling the command
```
`Then expect the integral part doubles the command`

efernandez replied:
or
`Then expect the integration do its part doubling the command`

efernandez commented on test/pid_tests.cpp:
```diff
@@ -152,11 +153,143 @@ TEST(ParameterTest, gainSettingCopyPIDTest)
+
+  // If call again with same arguments, no error change, but integration do its part
+  cmd = pid.computeCommand(-0.5, ros::Duration(1.0));
+  // Then expect command = 3x error againa
```
`again` (not `againa`)

efernandez commented on test/pid_tests.cpp:
```diff
@@ -152,11 +153,143 @@ TEST(ParameterTest, gainSettingCopyPIDTest)
+  // Expect the integral part to keep the previous command because ensure error = 0
+  EXPECT_EQ(-1.0, cmd);
+
+  // Double check that the integral contribution keep the previous command
```
`keeps`

efernandez commented on test/pid_tests.cpp:
```diff
@@ -152,11 +153,143 @@ TEST(ParameterTest, gainSettingCopyPIDTest)
+
+  // Call again with no error
+  cmd = pid.computeCommand(0.0, ros::Duration(1.0));
+  // Expect the integral part to keep the previous command because ensure error = 0
```
`it ensures`

efernandez commented on test/pid_tests.cpp:
```diff
@@ -152,11 +153,143 @@ TEST(ParameterTest, gainSettingCopyPIDTest)
+
+  // If call again doubling the error
+  cmd = pid.computeCommand(-1.0, ros::Duration(1.0));
+  // Then expect the command doubl-ed
```
`doubled`

efernandez commented:
LGTM, but I've added a few comments regarding typos on the comments. @bmagyar Maybe you can fix them before merging.

All the tests look good. Thanks for the contribution.

bmagyar commented:
Well I don't really have access to Carlos' branch, but I can fix it after I merged it! :)

Oh btw, merging...

bmagyar commented:
Oh shoot, should have rebased this one...

carlosjoserg commented:
Hey, sorry I'm late for the party... thanks for the review and the merge!

bmagyar commented:
No problem, thanks for your contribution!
